presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -470,6 +481,43 @@ public CounterStat getHighMemorySplitSource()
         return concat(partitionBatches);
     }
 
+    private static Optional<Set<HiveColumnHandle>> mergeRequestedAndPredicateColumns(Optional<Set<HiveColumnHandle>> requestedColumns, Set<HiveColumnHandle> predicateColumns)
+    {
+        if (!requestedColumns.isPresent() || predicateColumns.isEmpty()) {
+            return requestedColumns;
+        }
+
+        Map<String, Set<HiveColumnHandle>> nameToColumnHandles = new HashMap<>();
+        Stream.concat(requestedColumns.get().stream(), predicateColumns.stream())
+                .filter(column -> column.getColumnType() == REGULAR)
+                .forEach(column -> nameToColumnHandles.computeIfAbsent(column.getName(), (key) -> new HashSet<>()).add(column));
+
+        return Optional.of(nameToColumnHandles.values().stream()
+                .map(columnHandles -> {
+                    HiveColumnHandle finalHandle = null;
+                    for (HiveColumnHandle columnHandle : columnHandles) {
+                        if (finalHandle == null) {
+                            finalHandle = columnHandle;
+                            continue;
+                        }
+
+                        if (finalHandle.getHiveType().getCategory() != STRUCT || finalHandle.getRequiredSubfields().isEmpty()) {
```

Comment:
why can we short circuit for non-structs?  is it because there will never be any subfields?  I think the shortcircuiting is more confusing.

Reply:
Yes, because there will never be subfields in NON-structs. Well, there might be key, value stuff for maps, but that doesn't really impact encryption implementation since either the full map/array is encrypted or not.